File: stacks/0682-baseball-game.go
package stacks

// https://leetcode.com/problems/baseball-game/

// Approach: Optimized - Stack simulation
// Use stack for natural push/pop operations
// Process each operation: +, D, C, or number
// N = number of operations
// Time: O(N) - process each operation once, final sum is O(N)
// Space: O(N) - stack stores all valid scores
func calPoints(operations []string) int {
	stack := make([]int, 0)

	for _, op := range operations {
		n := len(stack)

		switch op {
		case "+":
			// Sum of previous two scores
			stack = append(stack, stack[n-1]+stack[n-2])
		case "D":
			// Double of previous score
			stack = append(stack, stack[n-1]*2)
		case "C":
			// Remove previous score
			stack = stack[:n-1]
		default:
			// It's a number (parse and add)
			var num int
			// Manual parsing to handle negative numbers
			sign := 1
			idx := 0

			if op[0] == '-' {
				sign = -1
				idx = 1
			}

			for ; idx < len(op); idx++ {
				num = num*10 + int(op[idx]-'0')
			}

			stack = append(stack, sign*num)
		}
	}

	// Calculate sum of all scores
	sum := 0
	for _, score := range stack {
		sum += score
	}

	return sum
}

File: stacks/0682-baseball-game.py
from typing import List
import unittest

# https://leetcode.com/problems/baseball-game/

# python3 -m unittest stacks/0682-baseball-game.py


class Solution(unittest.TestCase):
    # # Approach #1: Brute Force - List with operations
    # # Use a list to store all scores, perform operations on the list
    # # For each operation, append/remove from list, then sum at the end
    # # N = number of operations
    # # Time: O(N) - process each operation once
    # # Space: O(N) - list stores all valid scores
    # def calPoints(self, operations: List[str]) -> int:
    #     record: List[int] = []
    
    #     for op in operations:
    #         if op == "+":
    #             # Sum of previous two scores
    #             record.append(record[-1] + record[-2])
    #         elif op == "D":
    #             # Double of previous score
    #             record.append(record[-1] * 2)
    #         elif op == "C":
    #             # Remove previous score
    #             record.pop()
    #         else:
    #             # It's a number
    #             record.append(int(op))
    
    #     return sum(record)

    # Approach #2: Optimized - Stack simulation
    # Use stack for natural push/pop operations
    # Same logic but clearer intent with stack operations
    # N = number of operations
    # Time: O(N) - process each operation once, final sum is O(N)
    # Space: O(N) - stack stores all valid scores
    def calPoints(self, operations: List[str]) -> int:
        stack: List[int] = []

        for op in operations:
            if op == "+":
                # Sum of previous two scores
                stack.append(stack[-1] + stack[-2])
            elif op == "D":
                # Double of previous score
                stack.append(stack[-1] * 2)
            elif op == "C":
                # Remove previous score
                stack.pop()
            else:
                # It's a number (can be negative)
                stack.append(int(op))

        return sum(stack)

    def test(self):
        for operations, expected in [
            (["5", "2", "C", "D", "+"], 30),
            (["5", "-2", "4", "C", "D", "9", "+", "+"], 27),
            (["1", "C"], 0),
            (["1"], 1),
            (["1", "2", "+", "D", "C"], 6),
            (["-60", "D", "-36", "30", "13", "C", "C", "-33", "53", "79"], -117),
        ]:
            output = self.calPoints(operations)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")
